### PR TITLE
Move metrics handler from admin to health server

### PR DESCRIPTION
**Pull request by Jefftree**
admin server is binded only to localhost. This will allow metrics to be read by other containers that need to process the metrics.

/assign @cheftako 
/cc @caesarxuchao

### CODE REVIEW

**Comment by cheftako** [comment]
/assign @logicalhan 
Tagging in @logicalhan to make sure he has no issues with exposing metrics on a non localhost server.

File: cmd/server/main.go [hunk 1]
```diff
@@ -592,6 +591,7 @@ func (p *Proxy) runHealthServer(o *ProxyRunOptions, server *server.ProxyServer)
 	})
 
 	muxHandler := http.NewServeMux()
+	muxHandler.Handle("/metrics", promhttp.Handler())
 	muxHandler.HandleFunc("/healthz", livenessHandler)
 	muxHandler.HandleFunc("/ready", readinessHandler)
```
Comment (logicalhan) [inline cmd/server/main.go:596]: shouldn't this be `/readyz`?

File: cmd/server/main.go [hunk 2]
```diff
@@ -551,7 +551,6 @@ func redirectTo(to string) func(http.ResponseWriter, *http.Request) {
 
 func (p *Proxy) runAdminServer(o *ProxyRunOptions, server *server.ProxyServer) error {
 	muxHandler := http.NewServeMux()
-	muxHandler.Handle("/metrics", promhttp.Handler())
```
Comment (cheftako) [inline cmd/server/main.go:554]: Can we add a redirect (status code 301) handler to the new port? We can remove it in a bit (and I know we have not gone to 1.0 yet) but its good to start ingraining backward compatibility concerns.

Comment (logicalhan) [reply cmd/server/main.go:554]: Why not just leave this here for backwards compatibility?

Comment (cheftako) [reply cmd/server/main.go:554]: I'd be ok with that solution as well. I just don't want the endpoint to disappear without explanation or any attempt to maint compatibility.

File: cmd/server/main.go (hunk 1 above)
Comment (cheftako) [reply cmd/server/main.go:596]: shrug, Kubelet doesn't care. Happy to have it changed if we believe that is more standard.

File: cmd/server/main.go (hunk 2 above)
Comment (cheftako) [reply cmd/server/main.go:554]: If we are ever plan to stop serving /metrics on the admin port, then we should 301 to the new destination. If we're ok keeping it in perpetuity then go ahead and leave it. If we think we would rather just maintain it in one place then its probably better to put in the 301 now, before usage goes up.

**Comment by caesarxuchao** [comment]
Is it safe to expose metrics on non-localhost? Is it necessary?

**Comment by Jefftree** [comment]
Hmm, it seems that this has more implications than I originally expected. Will do some more investigation on alternatives before proceeding with this route.

/hold

**Comment by cheftako** [comment]
> Is it safe to expose metrics on non-localhost? Is it necessary?

Would love to get @logicalhan's (SIG Instrumentation) take on that.

**Comment by mikedanese** [comment]
Talked to Jeff and this isn't a publicly exposed endpoint (at least by default).

/lgtm

File: cmd/server/main.go (hunk 2 above)
Comment (Jefftree) [reply cmd/server/main.go]: Added the 301 redirect

**Comment by Jefftree** [comment]
/unhold

@cheftako @caesarxuchao PTAL, thanks!

File: cmd/agent/main.go [hunk 3]
```diff
@@ -257,7 +259,15 @@ func (a *Agent) runHealthServer(o *GrpcProxyAgentOptions) error {
...
 func (a *Agent) runAdminServer(o *GrpcProxyAgentOptions) error {
 	muxHandler := http.NewServeMux()
-	muxHandler.Handle("/metrics", promhttp.Handler())
+	muxHandler.Handle("/metrics", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		host, _, err := net.SplitHostPort(r.Host)
+		// The port number may be omitted if the admin server is running on port 80
+		if err != nil {
+			host = r.Host
+		}
+		http.Redirect(w, r, fmt.Sprintf("%s:%d%s", host, o.healthServerPort, r.URL.Path), 301)
```
Comment (caesarxuchao) [inline cmd/agent/main.go:268]: nit: s/301/http.StatusMovedPermanently

File: cmd/agent/main.go [hunk 4]
```diff
@@ -257,7 +259,15 @@ func (a *Agent) runHealthServer(o *GrpcProxyAgentOptions) error {
 
 func (a *Agent) runAdminServer(o *GrpcProxyAgentOptions) error {
 	muxHandler := http.NewServeMux()
-	muxHandler.Handle("/metrics", promhttp.Handler())
+	muxHandler.Handle("/metrics", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		host, _, err := net.SplitHostPort(r.Host)
+		// The port number may be omitted if the admin server is running on port 80
```
Comment (caesarxuchao) [inline cmd/agent/main.go:264]: how about "The port number may be omitted if the admin server is running on port 80, the default port for HTTP"?

**Pull request merged**